### ISSUE TITLE
Nightly channel does not see that new builds are installable

### ISSUE THREAD

**Issue by snomos** [divvun/divvun-manager-macos#32]
Taskcluster is working for all lang-xxx repos, and signing and notarization is working again. That is, we have now a bunch of new nightlies available to us. Except not:

<img width="870" alt="Skjermbilde 2022-03-09 kl  08 21 15-kopi" src="https://user-images.githubusercontent.com/1255285/157384786-06f94c5b-c594-4721-b5c0-658411753ec3.png">

Expected: available for installation

Divvun Manager 2.0.1.

**Comment by Eijebong** [comment]
Doesn't that show that it's already been updated?

**Comment by snomos** [comment]
Yes, that would be the expected behavior. But the installed speller is this:

<img width="545" alt="Skjermbilde 2022-03-09 kl  09 37 49" src="https://user-images.githubusercontent.com/1255285/157394588-34ec6fe8-bec9-4339-a8de-8e244eb29317.png">

That is, a more than one year old speller (I assume the latest stable release).

**Comment by Eijebong** [comment]
I wonder if the manager thinks it's the same version because 4.3.0 matches

**Comment by snomos** [comment]
I think that I was fooled by having installed the original package using the "stable" channel, which makes any newer releases in the nightly channel unavailable.

I will reformulate the bug as follows:

Original installation channel must be visible, updates listed accordingly.

**Comment by snomos** [comment]
More on this:

- as updates are installed based on the active channel when a package was first installed, it must be possible to see which channel is active for a given package
- listed updates should be listed based on the active channel (updates in other channels could be listed as well, in gray or something, to make it clear that they exist, but are not available via the active channel)

**Comment by snomos** [comment]
@dylanhand  Now that DM is building as it should again, and there is no further direct action to be taken on #41 (except adding logging), it would be good to take a look at this issue.

**Comment by dylanhand** [comment]
@snomos will do. I'll see if there are obvious places to add logging and then look into how we can indicate from which channel a package is installed.

**Comment by snomos** [comment]
I think it would be useful to have at least the following information directly readable in the list view in Divvun Manager:

- installed version (with active channel)
- we need column headers in the list view, so that we can tell users exactly what they are seeing, and also make the list view sortable by clicking the column headers (cf #44 )

**Comment by dylanhand** [comment]
> Taskcluster is working for all lang-xxx repos, and signing and notarization is working again. That is, we have now a bunch of new nightlies available to us. Except not:
> 
> <img alt="Skjermbilde 2022-03-09 kl 08 21 15-kopi" width="870" src="https://user-images.githubusercontent.com/1255285/157384786-06f94c5b-c594-4721-b5c0-658411753ec3.png">
> 
> Expected: available for installation
> 
> Divvun Manager 2.0.1.

The reason no updates are listed in this situation is because the nightly version number is equal to (or less than) the current stable release version.

**One solution is to bump the version number once a release has made it to the stable channel.** It could be argued that this is what should have been done all along since, once version 1.0.0 is released, you're no longer building nightly versions of 1.0.0, but rather nightly versions of forthcoming versions such as 1.0.1, 1.1.0, etc.

These steps were taken to test this:
1. Install Inari Sami spellchecker from the `stable` channel

<img width="569" alt="speller-smn 01 install from stable channel copy" src="https://github.com/divvun/divvun-manager-macos/assets/431529/2530fd0a-3754-4765-925a-af99bdd40295">


2. Switch to `nightly` channel, where current version is `1.0.5-nightly.20240517T030637371Z`. It shows "No Updates" because the version, 1.0.5, is the latest version, and it's installed (the current implementation of DM ignores the "`-nightly.2024...`" part.

<img width="735" alt="speller-smn 02 no updates listed before new nightly version bump copy" src="https://github.com/divvun/divvun-manager-macos/assets/431529/dd8429c4-2d87-460e-aea4-4046dcd36f34">


3. [Bump the speller patch version to produce a new nightly version](https://github.com/giellalt/lang-smn/commit/582f155344b80d3df4069f366afafb6ec2664a12), then refresh DM. This now shows that an update is available since 1.0.6 > 1.0.5:

<img width="732" alt="speller-smn 03 update available after bumping version number copy" src="https://github.com/divvun/divvun-manager-macos/assets/431529/638a4008-004f-4c6e-9a10-679e3d899c29">

---

#### TLDR

If you want nightly versions to be installable over stable versions, the easy fix that requires no changes is to bump the nightly's version number. The user will still have the option _not_ to update if they choose: they simply don't update when switched to the nightly channel.

---

This solution however **does not** address what this issue has morphed into, which is to display the channel a package was installed from.

So now, given this new information, the question is: is the current implementation acceptable (i.e., bump the nightly version number to make them available to install over stable versions)?

Let me know if you'd still like a column that lists the active channel from which a package is installed. As discussed in our meeting, I've worked on an experimental feature that lets us get the non-truncated package version (e.g. `1.0.0-nightly...` instead of just `1.0.0`) from the `Info.plist`, which would help us determine from which channel a package is installed.

One unknown is whether we can accurately determine whether a package was installed from the beta channel. Packages from the `nightly` channel always have "nightly" in their version string, so that's easy. If there's a similar requirement for beta packages (e.g. they always contain something like `-rc1`), that too will be easy. If not, we may have to figure out a way to identify packages installed from the beta channel.

**Comment by snomos** [comment]
@dylanhand before we can answer the question of whether we need an active channel column, we need to clarify the following:

when do automatic updates occur? The intended behaviour is that within the same channel, new releases should be installed automatically (the installed update will be available at latest after computer restart, due to macOS caching speller lexicons in memory, so an app restart might not be enough). Do automatic updates also occur _across_ channels if version numbers are changed as suggested above?

At least within one and the same channel, automatic updates do happen — the SMN update mentioned above got installed without me doing anything:

```sh
 ls -l /Library/Services/smn.bundle/Contents/Resources/smn.zhfst  
-rw-r--r--  1 root  wheel  10185021 May 20 16:42 /Library/Services/smn.bundle/Contents/Resources/smn.zhfst
echo nuvviDspeller | divvunspell suggest -a /Library/Services/smn.bundle/Contents/Resources/smn.zhfst
Reading from stdin...
Input: nuvviDspeller		[INCORRECT]
Divvun speller for Inari Sami		1
smn, desktop, version 1.0.6, 20.05.2024-1341		2
Built using HFST 3.16.0		3
```

I have now bumped the FAO speller in a similar manner; I have it installed from the stable channel ATM, let's see whether it is updated automatically to nightly if the truncated, semver version number is bumped.

The plan has all along been that automatic updates do occur as follows:
- always when new releases are available in the same channel as previously installed
- never across release channels

The confusion arises for people when they don't remember/see which channel a package was installed from, and expect it to be automatically updated, but then it does not.

**Comment by snomos** [comment]
Faroese update out (1.1.2-nightly), but not installed, and not listed as installable:

![image](https://github.com/divvun/divvun-manager-macos/assets/1255285/7399b45e-4f74-4bab-a22d-71c55dba6336)

I have this installed:

```sh
$ ls -l /Library/Services/*.bundle/Contents/Resources/*.zhfst  
-rw-r--r--  1 root  wheel   5276913 Oct  8  2021 /Library/Services/fo.bundle/Contents/Resources/fo.zhfst

$ echo nuvviDspeller | divvunspell suggest -a /Library/Services/fo.bundle/Contents/Resources/fo.zhfst 
Reading from stdin...
Input: nuvviDspeller		[INCORRECT]
Divvun speller for Faroese		1
fo, desktop, version 1.1.1, 08.10.2021-0722		2
Built using HFST 3.15.5		3
```

That is, for Faroese, the update is not installed automatically, and it is also not listed as manually installable, despite having a higher (truncated/semver) version number.

**Comment by snomos** [comment]
> That is, for Faroese, the update is not installed automatically, and it is also not listed as manually installable, despite having a higher (truncated/semver) version number.

I was too quick to draw conclusions. Now, an hour or so later, it has been automatically installed:

```sh
ll /Library/Services/*.bundle/Contents/Resources/*.zhfst                                             
-rw-r--r--  1 root  wheel   5336900 May 21 09:40 /Library/Services/fo.bundle/Contents/Resources/fo.zhfst
[...]

➜ echo nuvviDspeller | divvunspell suggest -a /Library/Services/fo.bundle/Contents/Resources/fo.zhfst
Reading from stdin...
Input: nuvviDspeller		[INCORRECT]
Divvun speller for Faroese		1
fo, desktop, version 1.1.2, 21.05.2024-0640		2
Built using HFST 3.16.0		3
```

That is, it seems you are right @dylanhand that updates happen irrespective of channel, and based on the truncated version number only.

**Comment by dylanhand** [comment]
Got it. This is what appears to have happened for me as well -- namely, the nightly Inari Sami spellchecker (1.0.6-nightly) was installed over the stable version (1.0.5).

So then it sounds like the requirements are:
1. Divvun Manager (and the RPC) need to be updated to keep track of which channel a package was installed from
2. That channel should be listed in the DM UI
3. Ensure automatic updates only update packages from the same channel from which a package was initially installed

I know it will be possible to differentiate stable/nightly versions since nightly versions always have "nightly" in the version number.

Is there a similar convention for beta versions? I see that some are of the form `0.1.1-rc1`. Will this always be the case?

If not we'll need a different way of identifying packages installed via the beta channel.

![CleanShot 2024-05-21 at 10 53 18@2x](https://github.com/divvun/divvun-manager-macos/assets/431529/c49110b5-b651-4f9f-b427-5230e85e1013)

**Comment by snomos** [comment]
> So then it sounds like the requirements are:
>
> 1. Divvun Manager (and the RPC) need to be updated to keep track of which channel a package was installed from
> 1. That channel should be listed in the DM UI
> 1. Ensure automatic updates only update packages from the same channel from which a package was initially installed

Yes.

The nightly/beta/regular release algorithm is very simple:

- nightly: contains `nightly` in the version string
- beta: does _not_ contain `nightly`, and version number is `< 1.0.0`
- standard release: does _not_ contain `nightly`, and version number is `≥ 1.0.0`

That is, we don't do betas after the first public, full release (the 1.0.0 release).

**Comment by dylanhand** [comment]
Ok great, thanks for clarifying the release details. I'll implement it that way.

**Comment by dylanhand** [comment]
FYI, here's a mock-up of what I have in mind for listing used channel. Packages that are installed will show what channel they were installed from. Packages that are not installed show nothing in this column.

![channel-list](https://github.com/divvun/divvun-manager-macos/assets/431529/6ec8231a-440e-4cf8-b41f-65b9d2947a56)


Please let me know if you agree or see it differently 🙂

**Comment by snomos** [comment]
Looks good, but also amplifies the need for column headers. What do these labels mean?

**Comment by dylanhand** [comment]
Agreed. I'll look into adding column headers at the same time.

**Comment by dylanhand** [comment]
As discussed offline, this is a wontfix.

Divvun Manager was built to have one channel for everything. The main point of confusion was the need to bump nightly version numbers for them to supersede stable versions.

Changes to Pahkat have been made that fix #48.

@snomos can you close this issue if you agree?

**Comment by dylanhand** [comment]
Thanks for the info - I'll look into what's causing `lang-kal` to behave differently from the others.

**Comment by dylanhand** [comment]
I was unable to reproduce the issue with `lang-kal`.

These were my steps:
1. Install the [latest nightly](https://pahkat.uit.no/artifacts/divvun-installer_2.2.0-nightly.20240527T120810037Z_macos.pkg) of Divvun Manager Mac
2. Restart
3. Use DM to install the speller for `lang-kal`
4. Restart
5. Take note of the installed version (`1.0.1-nightly.20240530T215916935Z`)
6. Deploy new version of `lang-kal`
7. Go to lunch
8. Force-refresh DM UI by switching from "All Repositories" to "Divvun Languages", and back again

Result:
The new speller was installed while I was away at lunch.

Running your command from above also shows that the `kl.zhfst` file was updated:
```
ls -l /Library/Services/*.bundle/Contents/Resources/*.zhfst
.rw-r--r-- root wheel  25 KB Thu Feb  1 09:58:43 2024  /Library/Services/aka.bundle/Contents/Resources/ak.zhfst
...
.rw-r--r-- root wheel  18 MB Fri May 31 12:21:09 2024  /Library/Services/kl.bundle/Contents/Resources/kl.zhfst
...
.rw-r--r-- root wheel  34 KB Thu Feb  1 11:43:42 2024  /Library/Services/zul-x-exp.bundle/Contents/Resources/zu.zhfst

```

**Comment by snomos** [comment]
I notice an inconsistency in the file listing above (last comment) regarding the bundle names:

- aka and zul is using the full BCP-47 tag including ISO 639-3 codes in the bundle names (ie the same as used for the repository name)
- kl is using the ISO 639-1 code both for bundle name and zhfst file

Could this inconsistency (ie not the same bundle naming conventions everywhere) be causing trouble for DM in finding or detecting the bundle and the bundle data/timestamp?

It makes sense that the bundle name should be the full BCP-47 string from the repo name, as that makes it possible to have parallel installs for the same language but different repos installed at the same time. There are several languages with multiple (two) repos, for various reasons, and being able to install spellers from them all/both makes it possible to test both spellers or choose which one to use.
